src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc

```diff
@@ -523,17 +524,21 @@ WeightedRoundRobin::PickResult WeightedRoundRobin::Picker::Pick(
 }
 
 size_t WeightedRoundRobin::Picker::PickIndex() {
-  // Grab the mutex and do WRR pick if we have a scheduler.
+  // Grab a ref to the scheduler.
+  std::shared_ptr<StaticStrideScheduler> scheduler;
   {
     MutexLock lock(&mu_);
-    if (scheduler_.has_value()) {
-      return scheduler_->Pick();
-    }
+    scheduler = scheduler_;
   }
+  // If we have a scheduler, use it to do a WRR pick.
+  if (scheduler != nullptr) return scheduler->Pick();
   // We don't have a scheduler (i.e., either all of the weights are 0 or
   // there is only one subchannel), so fall back to RR.
-  last_picked_index_ = (last_picked_index_ + 1) % subchannels_.size();
-  return last_picked_index_;
+  size_t index = last_picked_index_.fetch_add(1);
+  if (index > 0 && index % subchannels_.size() == 0) {
+    last_picked_index_.fetch_sub(subchannels_.size());
```

Comment:
Do we need this? I think we can let it just grow and wrap and always take the modulo.

Reply:
I think it's not critical but still not sure if we need this. When it wraps it will jump to an index which will be approximately uniformly random, which is the same as when we create a new picker. Given we are not concerned with the impact of a new picker that much I think it's fine to not have this. And when the subchannel list is very small I guess the overhead becomes a bit unnecessary.

Reply:
We create a new picker only when something is changing (either the list of addresses changes or a subchannel connectivity state changes), in which case there's already going to be churn.  In the case where there is no external churn, I don't think we want this kind of discontinuity.

The existing RR policy in gRPC already does this, so I think it's good to be consistent here -- and this code path will be executed only when we don't have any weights anyway.  But I'll follow up on whether or not this is worth doing performance-wise.

Reply:
After discussing this with the other languages, it looks like this isn't really necessary, so I've removed it.